gen/internals.py

```diff
@@ -0,0 +1,464 @@
+
+
+class Scope:
+    def __init__(self, name: str, cases=None):
+        self.name = name
+        self.cases = cases if cases else dict()
+
+    def add_case(self, value: str, target):
+        assert isinstance(target, Target)
+        self.cases[value] = target
+
+    def __iadd__(self, other):
+        assert isinstance(other, Scope), "Internal consistency error, expected Scope but got {}".format(type(other))
+
+        # Must have the same name and same options in order to be merged (can't have a new
+        # switch add new, unhandled cases to a switch already included / represented by this scope).
+        assert self.name == other.name, "Internal consistency error: Trying to merge scopes with " \
+            "different names: {} and {}".format(self.name, other.name)
+        assert self.cases.keys() == other.cases.keys(), "Same name / switch variable introduced " \
+            "with a different set of possible cases. name: {}. First options: {}, Second " \
+            "options: {}".format(self.name, self.cases.keys(), other.cases.keys())
+
+        # Merge the targets for the cases
+        for name in self.cases:
+            self.cases[name] += other.cases[name]
+
+        return self
+
+    def __eq__(self, other):
+        assert isinstance(other, Scope)
+        return self.name == other.name and self.cases == other.cases
+
+
+class Target:
+
+    # TODO(cmaloney): Make a better API for working with and managing sub scopes. The current
+    # dictionary of dictionaries is really hard to use right.
+    def __init__(self, variables=None, sub_scopes=None):
+        self.variables = variables if variables else set()
+        self.sub_scopes = sub_scopes if sub_scopes else dict()
+
+    def add_variable(self, variable: str):
+        self.variables.add(variable)
+
+    def add_scope(self, scope: Scope):
+        self.sub_scopes[scope.name] = scope
+
+    def __iadd__(self, other):
+        assert isinstance(other, Target), "Internal consistency error, expected Target but got {}".format(type(other))
+        self.variables |= other.variables
+
+        # Add all scopes from the other to this, merging all common sub scopes.
+        for name, scope in other.sub_scopes.items():
+            if name in self.sub_scopes:
+                self.sub_scopes[name] += scope
+            else:
+                self.sub_scopes[name] = scope
+
+        return self
+
+    def __eq__(self, other):
+        assert isinstance(other, Target)
+        return self.variables == other.variables and self.sub_scopes == other.sub_scopes
+
+
+class Source:
+    def __init__(self, entry=None, is_user=False,):
+        self.setters = dict()
+        self.validate = list()
+        self.is_user = is_user
+        if entry:
+            self.add_entry(entry, False)
+
+    def add_setter(self, name, value, is_optional, conditions):
+        self.setters.setdefault(name, list()).append(Setter(name, value, is_optional, conditions, self.is_user))
```

Comment:
this isn't doing anything like that with it, it's just saying `if the dictionary setters has an entry for name already do nothing, otherwise create an entry and set it to be an empty list`. We then append to that list. setdefault() is doing an "initialize if uninitialized".